Display transcript text and follow along the audio

tonytamsf opened the pull request:
fix #4935

### Description
Display contents of podcast:transcript tag as a full dialog view using RecylerView.
1. When tapping the currently playing transcript paragraph, tapping it will pause or start the podcast
2. When tapping another paragraph, will skip and start playing
3. better handling of scrolling, so the user can scroll to future and past and not have the software automatically scroll back to the current stat of the audio

### Yet to do
- allow for copy and pasting
- allow for sharing and episode at a certain time from the transcript

### Future
- allow for searching for text in the transcript


### Image & Video

## Version 0.9
Follow audio checkbox


https://github.com/AntennaPod/AntennaPod/assets/149837/cc55a7dc-cdbe-4396-8e8b-e7e94ce3fa8d


![follow audio](https://github.com/AntennaPod/AntennaPod/assets/149837/00e71df4-02a3-47eb-bf1b-9cb83e18ec45)


## Version 0.1
<img width="417" alt="Screenshot 2024-04-14 at 10 50 34 PM" src="https://github.com/AntennaPod/AntennaPod/assets/149837/c4e0008f-c11f-4bbe-996d-686d336080df">


https://github.com/AntennaPod/AntennaPod/assets/149837/eee2ac20-9eb3-4ad5-a079-c473e45d3e95

Matth7878 commented:
Nice. A suggestion after looking at the screenshot. I think text should be justified to make it easier to eyes.

keunes commented:
@Matth7878 Full justified text is only acceptable when hyphenation is applied correctly. Otherwise you get 'rivers' (vertical lanes of white space) which aren't great, especially for folks with dyslexia. I don't know if good hyphenation (which of course is langue-dependant) is doable in Android

Some resources on this:
https://blog.prototypr.io/text-alignment-best-practises-c4114daf1a9b
https://yesimadesigner.com/justification-vs-alignment/
https://uxmovement.com/content/6-surprising-bad-practices-that-hurt-dyslexic-users/

keunes commented:
@tonytamsf Nice! I think I'd put it in a full window, though, instead of a modal. Gives just that little extra space, and with a top bar we get a logical place to add a search icon (in future).

tonytamsf commented:
> @tonytamsf Nice! I think I'd put it in a full window, though, instead of a modal. Gives just that little extra space, and with a top bar we get a logical place to add a search icon (in future).

@keunes 
If we did the full window, do we have to handle any of the player controls like forward, pause?

Thoughts about dialog vs full fragment view @ByteHamster ?

keunes commented:
Hi @tonytamsf

> If we did the full window, do we have to handle any of the player controls like forward, pause?

Not in any different way from a dialog, I would say. (I believe we did discuss having player controls but potentially in future.)

Maybe a full screen dialog would be easier to implement than a dedicated fragment?

https://m3.material.io/components/dialogs/specs

ByteHamster commented on ui/chapters/src/main/java/de/danoeh/antennapod/ui/chapters/PodcastIndexTranscriptUtils.java:
```diff
@@ -2,32 +2,44 @@
 
     private static final String TAG = "PodcastIndexTranscript";
 
+
```
Please revert white space change. Might also be caused by my rebasing, not sure

tonytamsf replied:
remove whitespace

ByteHamster commented on ui/chapters/src/main/java/de/danoeh/antennapod/ui/chapters/PodcastIndexTranscriptUtils.java:
```diff
@@ -2,32 +2,44 @@
                 str.append(response.body().string());
             }
-        } catch (IOException e) {
+        } catch (Exception e) {
```
I replaced this because it prevented downloading files when the transcript code failed (in my case, because the transcript url was invalid)

ByteHamster commented on ui/chapters/src/main/java/de/danoeh/antennapod/ui/chapters/PodcastIndexTranscriptUtils.java:
```diff
@@ -37,6 +49,48 @@ public static String loadTranscriptFromUrl(String type, String url, boolean forc
+            }
+        }
+
+        String transcriptUrl = media.getItem().getPodcastIndexTranscriptUrl();
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        Future<Transcript> future = executor.submit(() -> {
+            String t = PodcastIndexTranscriptUtils.loadTranscriptFromUrl(transcriptType, transcriptUrl, false);
+            if (StringUtils.isNotEmpty(t)) {
+                return TranscriptParser.parse(t, transcriptType);
+            }
+            return null;
+        });
+
+        try {
+            Transcript result = future.get();  // This will block until the Callable completes
+            executor.shutdown();  // Remember to shutdown the executor
+            return result;
+        } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
+        }
```
This looks like a workaround for `NetworkOnMainThreadException`. While it hides the exception, it still blocks the main thread until the download is done. Instead, the transcript should be loaded in the background without blocking the main thread. See `ChaptersFragment.loadMediaInfo`

ByteHamster commented on ui/common/src/main/res/drawable/subtitles_24px.xml:
```diff
@@ -0,0 +1,10 @@
+    android:height="24dp"
+    android:viewportWidth="960"
+    android:viewportHeight="960"
+    android:tint="?attr/colorControlNormal">
+  <path
+      android:fillColor="@android:color/white"
```
For other action icons we use `?attr/action_icon_color` directly and don't have to tint then

ByteHamster commented on ui/i18n/src/main/res/values/strings.xml:
```diff
@@ -854,4 +855,5 @@
     <string name="shortcut_subscription_label">Subscription shortcut</string>
     <string name="shortcut_select_subscription">Select subscription</string>
     <string name="add_shortcut">Add shortcut</string>
+    <string name="transcript_label">Transcripts</string>
```
I think this should be singular. Not a native speaker, though

ByteHamster reviewed (commented):
Thanks! It's nice seeing the transcript run by on by own phone. It already works great.

I added a bunch of comments below. The CI server also left a comment about a SpotBugs violation (that's new with the rebase :) )

tonytamsf commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/ItemTranscriptRvAdapter.java:
```diff
@@ -0,0 +1,159 @@
+ * {@link RecyclerView.Adapter} that can display a {@link PlaceholderItem}.
+ * TODO: Replace the implementation with code for your data type.
+ */
+public class ItemTranscriptRvAdapter extends RecyclerView.Adapter<ItemTranscriptRvAdapter.ViewHolder> {
```
renamed to TranscriptAdapter

tonytamsf commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/ItemTranscriptRvAdapter.java:
```diff
@@ -0,0 +1,159 @@
+        return transcript.getSegmentsMap().size();
+    }
+
+    public class ViewHolder extends RecyclerView.ViewHolder {
```
moved file

tonytamsf commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/ItemTranscriptRvAdapter.java:
```diff
@@ -0,0 +1,159 @@
+        if (transcript == null) {
+            return;
+        }
+        TreeMap<Long, TranscriptSegment> segmentsMap = transcript.getSegmentsMap();
```
The TreeMap does not have a function to find the position (index) of an entry.  I've moved this into the `Transcript` class instead into a `getIndex()` function

ByteHamster replied:
Hmm, if you need the index, maybe a TreeMap is then not the right data structure? Maybe a simple ArrayList with binary search would be a better fit then? Then we can't end up in a state where both data structures somehow have conflicting data. In terms of access time, both binary search and TreeMap should have log(n). I think a TreeMap only has performance advantages when the data is dynamic.

tonytamsf replied:
I was going to move the internal implementations into Transcript and not expose TreeMap to the outside of the class, maybe that will help a bit with the readability.

tonytamsf commented on app/src/main/res/layout/fragment_item_transcript_rv_list.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
```
renamed to transcript_dialog.xml

tonytamsf commented on app/src/main/res/layout/fragment_item_transcript_rv.xml:
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8"?>
```
renamed to `transcript_item.xml`

tonytamsf commented on parser/transcript/src/main/java/de/danoeh/antennapod/parser/transcript/TranscriptParser.java:
```diff
@@ -1,11 +1,12 @@
 
 public class TranscriptParser {
-    static final long MIN_SPAN = 1000L; // merge short segments together to form a span of 1 second
+    static final long MIN_SPAN = 5000L; // merge short segments together to form a span of 1 second
```
fixed comment

tonytamsf commented on parser/transcript/src/main/java/de/danoeh/antennapod/parser/transcript/TranscriptParser.java:
```diff
@@ -21,4 +22,12 @@ public static Transcript parse(String str, String type) {
         return null;
     }
+
+    public static String secondsToTime(long msecs) {
```
using `Converter.getDurationStringLong`

tonytamsf reviewed (commented):
address most comments, still have a few more to go

tonytamsf commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptFragment.java:
```diff
@@ -0,0 +1,240 @@
+                    ((TextView) prevView.findViewById(R.id.content)).setTypeface(null, Typeface.NORMAL);
+                    ((TextView) prevView.findViewById(R.id.content)).setTextColor(
+                            ThemeUtils.getColorFromAttr(getContext(), android.R.attr.textColorSecondary));
+                }
```
I'm doing highlighting in one row and removing the highlight in the other row.  You think I should still move the code to the adapter?

cc @ByteHamster

ByteHamster replied:
I would say so, yes. The rows in the adapter can then decide themselves whether they should be highlighted. The fragment can maybe remember which segment was highlighted last and then notify changes for both the new one and the old one. This ensures that when scrolling quickly, the view states get handled properly.

tonytamsf commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptFragment.java:
```diff
@@ -0,0 +1,290 @@
+        params = getDialog().getWindow().getAttributes();
+        params.width = WindowManager.LayoutParams.MATCH_PARENT;
+        getDialog().getWindow().setAttributes((WindowManager.LayoutParams) params);
+        getDialog().getWindow().setBackgroundDrawable(new ColorDrawable(Color.BLACK));
```
@ByteHamster - any tip on how to get the background color for the theme?

this didn't work

```
        getDialog().getWindow().setBackgroundDrawableResource(ThemeUtils.getDrawableFromAttr(getActivity(),
                R.attr.selectableItemBackground)
        );```

ByteHamster replied:
Hmm, it feels a bit strange that this needs to be set manually. Can't you just remove setting the window background? Then the Material3 library handles the color.

tonytamsf replied:
removed the need for custom setting, it was because of the `AppCompatDialogFragment`subclass (which I removed)

tonytamsf commented:
@ByteHamster I have 2 questions for you in the PR.

Address all of your other review comments

tonytamsf commented on ui/chapters/src/main/java/de/danoeh/antennapod/ui/chapters/PodcastIndexTranscriptUtils.java:
```diff
@@ -2,32 +2,44 @@
             Log.d(TAG, "Downloading transcript URL " + url.toString());
-            Request request = new Request.Builder().url(url).build();
+            Request request = new Request.Builder().url(url).cacheControl(CacheControl.FORCE_NETWORK).build();
             response = AntennapodHttpClient.getHttpClient().newCall(request).execute();
```
Not sure why this doesn't work consistently when trying to download transcripts while streaming.
cc @ByteHamster

ByteHamster replied:
Hmm, it does seem to work for me. Does it print an error message or what is happening there?

tonytamsf reviewed (commented):
Left @ByteHamster a question about why the streaming of Transcript doesn't always work ui/chapters/src/main/java/de/danoeh/antennapod/ui/chapters/PodcastIndexTranscriptUtils.java

tonytamsf reviewed (commented):
Left @ByteHamster a question about why the streaming of Transcript doesn't always work 

`ui/chapters/src/main/java/de/danoeh/antennapod/ui/chapters/PodcastIndexTranscriptUtils.java`

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptAdapter.java:
```diff
@@ -0,0 +1,126 @@
+        return media.getTranscript().getSegmentsMap().size();
+    }
+
+    public interface Callback {
```
Maybe `SegmentClickListener`? Callback could be anything

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptFragment.java:
```diff
@@ -0,0 +1,290 @@
+    TranscriptViewholder prevView = null;
+    RecyclerView viewTranscript;
+
+    @Override
+    public void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+    }
```
Can be deleted, does not contain content

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptFragment.java:
```diff
@@ -0,0 +1,290 @@
+    Playable media;
+    Transcript transcript;
+    TreeMap<Long, TranscriptSegment> segmentsMap;
+    private ProgressBar progressBar;
```
Can be accessed through `ViewBinding` instead, no need to keep the additional reference here

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptFragment.java:
```diff
@@ -0,0 +1,290 @@
+    TranscriptAdapter adapter = null;
+    TranscriptViewholder currentView = null;
+    TranscriptViewholder prevView = null;
+    RecyclerView viewTranscript;
```
Can be accessed through `ViewBinding` instead, no need to keep the additional reference here

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptFragment.java:
```diff
@@ -0,0 +1,290 @@
+                .setTitle(R.string.transcript)
+                .create();
+        dialog.show();
+        dialog.getButton(DialogInterface.BUTTON_NEUTRAL).setVisibility(View.INVISIBLE);
```
Should the neutral button have a text then? Currently it is invisible but can be clicked

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptFragment.java:
```diff
@@ -0,0 +1,290 @@
+        loadMediaInfo(true);
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+    }
```
Can be removed

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptFragment.java:
```diff
@@ -0,0 +1,290 @@
+        super.onPause();
+    }
+
+    public void scrollToPlayPosition(long playPosition) {
```
Maybe an idea for the automatic "following" of the scroll position could be: If the previously active segment is not visible, the user must have scrolled away, so don't scroll to the new one. Maybe not perfect but I think it is better than always scrolling

tonytamsf replied:
I added a 'Follow audio' checkbox, which will only appear when the user manually scrolls.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptFragment.java:
```diff
@@ -0,0 +1,290 @@
+    }
+
+    public void scrollToPosition(Integer pos) {
+        RecyclerView rv = viewBinding.transcriptList;
+        if (rv == null) {
+            return;
+        }
```
I don't think this check is needed. ViewBinding ensures at compile time that the views exist

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/TranscriptFragment.java:
```diff
@@ -0,0 +1,290 @@
+            rv.getLayoutManager().startSmoothScroll(smoothScroller);
+        }
+
+        RecyclerView.LayoutManager lm = rv.getLayoutManager();
+        if (lm == null) {
+            return;
+        }
+        View v = lm.findViewByPosition(pos);
+        if (v == null) {
+            return;
+        }
```
This feels a bit too careful :) So many checks make the code harder to read

tonytamsf replied:
removed most of the useless checks

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/audio/AudioPlayerFragment.java:
```diff
@@ -271,6 +271,7 @@ private void loadMediaInfo(boolean includingChapters) {
             if (media.getChapters() == null && !includingChapters) {
                 loadMediaInfo(true);
             }
+            // TODO should we load transcripts here?
```
I don't think so. The chapters are only loaded here because we show them on the seek bar. We don't show the transcripts here, so we don't need to load them

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/ui/screen/playback/audio/AudioPlayerFragment.java:
```diff
@@ -495,6 +496,11 @@ public boolean onMenuItemClick(MenuItem item) {
             new SleepTimerDialog().show(getChildFragmentManager(), "SleepTimerDialog");
             return true;
+        } else if (itemId == R.id.transcript_item) {
+            new TranscriptFragment()
+                    .initMedia(media)
```
This will break when the Android system creates a new instance of the fragment. Instead, the arguments should be passed with `setArguments`. For most other fragments, we have a method `newInstance` that handles this. See for example `FeedItemFilterDialog.newInstance`

ByteHamster commented on app/src/main/res/layout/transcript_dialog.xml:
```diff
@@ -0,0 +1,26 @@
+        android:name="de.danoeh.antennapod.ItemTranscriptRVFragment"
+        android:layout_marginLeft="16dp"
+        android:layout_marginRight="16dp"
+        tools:context="de.danoeh.antennapod.fragment.ItemTranscriptRVFragment"
```
Outdated name?

ByteHamster commented on app/src/main/res/layout/transcript_dialog.xml:
```diff
@@ -0,0 +1,26 @@
+        android:id="@+id/transcript_list"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:name="de.danoeh.antennapod.ItemTranscriptRVFragment"
```
What does this do?

ByteHamster commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,7 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.6-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.1.1-bin.zip
```
Is the gradle update necessary? I would prefer to keep such changes separate in case they need to be reverted

ByteHamster commented on model/src/main/java/de/danoeh/antennapod/model/feed/Transcript.java:
```diff
@@ -1,23 +1,69 @@
 
 public class Transcript {
 
     private final TreeMap<Long, TranscriptSegment> segmentsMap = new TreeMap<>();
+    private Set<String> speakers;
+    private Hashtable<Long, Integer> positions;
+    private Object[] objs; // array of TreeMap<Long, TranscriptSegment>
```
Having TreeMap and Hash table and array somehow feels a bit duplicated. Just a simple array should have the same performance and would be a lot easier to comprehend when reading this code later

ByteHamster reviewed (commented):
Nice, this is making great progress!

tonytamsf reviewed (commented):
I did most of the changes, except swapping out `TreeMap`

ByteHamster commented:
Sorry for breaking checkstyle. It got later in the evening than I expected and I just wanted to get it pushed

tonytamsf commented:
> Sorry for breaking checkstyle. It got later in the evening than I expected and I just wanted to get it pushed 

Hans Peter, for transcript PR, you want me to change the TreeMap data structure? That will take me probably 1.5 weeks to refactor the code since I am in France.  Could I do a fast follow PR instead?  That way we can get this feature merged into transcript branch and I will refactor after merge to develop

keunes commented:
> Could I do a fast follow PR instead?

You mean [guid-based](https://podcasting2.org/podcast-namespace/tags/guid) fast-follow? That would be cool. Maybe we should create a separate issue to discuss the details?

tonytamsf commented:
> > Could I do a fast follow PR instead?
> 
> You mean [guid-based](https://podcasting2.org/podcast-namespace/tags/guid) fast-follow? That would be cool. Maybe we should create a separate issue to discuss the details?

Haa, nope. I meant I can follow up with another PR after this PR is merged to fix the internal data structure.  This way we can get people testing transcript while I clean up the code.

keunes commented:
Ahh, ok. Was worth a try :P

ByteHamster commented:
I just pushed a change to use a simple array with binary search. Seems to work fine and reduces the number of different data structures significantly.

I think something broke recently (between my previous commit and [before] the current one). The transcripts are concatenated for much longer sequences than just 5 seconds. To be exact, it seems that all text of one speaker is in one single segment

ByteHamster commented on parser/transcript/src/main/java/de/danoeh/antennapod/parser/transcript/JsonTranscriptParser.java:
```diff
@@ -33,13 +50,37 @@ public static Transcript parse(String jsonStr) {
+                segmentBody += body;
+                if (duration >= TranscriptParser.MIN_SPAN
+                        && ((i + 1) <= objSegments.length()
+                        && StringUtils.isAlphanumeric(
```
This alphanumeric check seems to prevent any merging of segments as soon as they use punctuation or similar

ByteHamster replied:
I think what you should check is if the segment *starts* with an alphanumeric character. Not if consists entirely of alphanumeric characters. Also, there should probably be a limit. If a segment takes longer than, say, 8 seconds, it should not be merged -- even if that means it will start with a non-alphanumeric character. I was testing with this podcast: https://logbuch-netzpolitik.de/abonnieren

tonytamsf replied:
fixed, thanks for the test case

ByteHamster commented:
Thanks for adding the "follow audio" checkbox, that's nice to have!

I think there are some small quirks left. Would be great if you could have a look.

- When scrolling manually, the checkbox stays checked and it jumps back to the playback position. I think the box should automatically un-check when scrolling manually.
- Sometimes the dialog remains empty. I can reproduce this when turning off the network connection but I also had it happen with network connection.
- When there are many 1-line segments, the currently playing one is sometimes cut off on top. Maybe that happens because scrolling to the `position-1` has some kind of offset. Maybe there is a way to scroll to the first *visible* position? Not sure.
- Sometimes when auto-scrolling, it keeps the playing episode below the screen, not on top of it. On my device this mainly happens when manually scrolling to the top, but it also happens when freshly opening the screen
- The logs print a bunch of `InterruptedIOException`s. These are harmless but probably confuse future contributors (we had something similar with the chapter screen). I would just print a single log line for `InterruptedIOException`s instead of a full stack trace.
- The `forceRefresh` parameter is unused, it always does a force-refresh. I think it should only do a force-refresh when clicking the corresponding button. Otherwise it cannot be viewed offline, for example. I think it should use the same logic as the online chapters

Showcasing the scrolling below the screen and the cut-off first episode: https://drive.google.com/file/d/1-ik8dUhYOB_3rEmiMwNnhLB6YZ40BAOY/view?usp=sharing

keunes commented:
Very cool! Some UI comments:
* When highlighting, I wouldn't highlight the timestamps and speakers. They're a bit distracting.
* When changing highlights, the animations (colour changes) can be a wee bit faster. It now sometimes feels like the highlights are lagging a bit, which I think is just an impression.
* The text is a bit small to read, seems smaller than the podcast description text. I would align the two.
* When there's speaker names, maybe we could extend the maximum segment time? I initially thought that maybe we should collate everything into one segment as long as it's the same speaker. That would work nice for the Podcasting 2.0 podcast (as Adam and Dave frequently hand over the speaking stick), but I think in other contexts the segments would then become too long (e.g. long-form interviews). Not sure what the time limit should be.
* The 'Follow audio' toggle doesn't display initially, only when you start scrolling (is my impression). I would just display it always (especially when turning off automatically as suggested by @ByteHamster).


Side issue: the automatic transcripts of Podcasting 2.0 are… not great ^^ Hope the speech recognition tech will improve further (or people edit manually before releasing).

tonytamsf commented:
@ByteHamster and @keunes Thanks for testing and the feedback, I'll work on changes to fix.  One thing I'm struggling with is the checkbox and the Recylerview doesn't seem to work well together inside a layout inside of the dialog.  The RecyclerView seems to overlap with the checkbox even thought I'm using a LinearLayout.  If you have any ideas, please let me know.

The other feedback, I'll track down and fix probably in about 1 week.

tonytamsf commented:
@keunes and @ByteHamster - could you test this for transcript again.  I think I have addressed all your comments. 

Also @ByteHamster can you help me move this file `de.danoeh.antennapod.ui.chapters.PodcastIndexTranscriptUtils` to another module?  It does not fit in `ui.chapters`, but I failed to move it to another module and fix the dependencies

keunes commented:
Thanks @tonytamsf!

A few additional comments/thoughts:
* For me the tickbox seems to not untick automatically on manual scroll.
* The current segment highlight is better. Just one additional tweak I would like to request: tiny left and right padding (there already seems a top & bottom padding, probably thanks to the line height).
* When tapping a segment, the timestamps & names are highlighted still. I would change that to only cover the text itself as well.
* When currently paused, then tapping on a segment initiated playback. I think it would be better to jump position but stay paused (e.g. if you want to share the episode with relevant timestamp).
* The toggle, refresh and close buttons are not vertically centred in the available space and have a rather small bottom margin.
* I also had the transcript screen staying empty (as @ByteHamster mentioned), after switching to another app and back, while episode was paused and transcript dialog was shown. Interestingly, tapping the 'close' button made the empty dialog go away and display immediately the (previously rendered?) dialog _with_ text.
* I'm still wondering if we shouldn't make this a full-screen dialog. Although the current form has been growing on me.
* I would increase the font size a bit and align it with what we have for the episode description.

ByteHamster commented:
There is still one quirk I don't understand. When opening the dialog for the second time while playing, it crashes. Error message: `Inconsistency detected. Invalid item position`. The chapters fragment does pretty much the same thing and does not crash. :thinking:

tonytamsf commented:
> There is still one quirk I don't understand. When opening the dialog for the second time while playing, it crashes. Error message: `Inconsistency detected. Invalid item position`. The chapters fragment does pretty much the same thing and does not crash. :thinking: 

@ByteHamster Remind me of which podcast you are seeing this?  Using Podcasting 2.0, I am not seeing this problem.

ByteHamster commented:
I think I fixed this in [3d09a7c](https://github.com/AntennaPod/AntennaPod/pull/7103/commits/3d09a7c6f178b416b76d1ecc20f85a2174aec821). The problem was that when loading was interrupted, it returned `null` instead of throwing. This meant that it re-tried loading the chapters (with force network) in an already interrupted thread, which caused a race condition with updating the RecyclerView.

ByteHamster commented:
Looks ready for merging to me. Is there anything open from your side, tony?

tonytamsf commented:
> Looks ready for merging to me. Is there anything open from your side, tony?

I have one more fix to make the dialog buttons not scroll and disable the refresh button during a refresh.

tonytamsf commented:
> Looks ready for merging to me. Is there anything open from your side, tony?

I'm ready to merge @ByteHamster now, the refresh button is disabled during a refresh

- @keunes I think the font size is the same in the dialog compare with the description
- I see @ByteHamster moved the transcript button to the menu, I'm fine with that during the initial launch, it's hard to discover it though

tonytamsf commented:
> Thanks @tonytamsf!
> 
> A few additional comments/thoughts:

We have addressed all of your comments except the font size, I think they are the same.

tonytamsf commented:
@keunes that crash looks unrelated, could you tell me which SHA is the AP debug build you are running?

I suspect it is not the right build because the overflow button for transcript is not there

keunes commented:
Hi @tonytamsf
3.3.2 (37ef84f). I just downloaded the debug build from GH right before testing.

ByteHamster commented:
> I see @ByteHamster moved the transcript button to the menu, I'm fine with that during the initial launch, it's hard to discover it though

Yeah, that's because we already have too many toolbar icons for small screens. Having it in the overflow menu is indeed hard to discover but I wanted to merge your PR before we start working on redesigning the player screen

> 3.3.2 (https://github.com/AntennaPod/AntennaPod/commit/37ef84f09cd4e1e0c7373774758bc04617e6eb60). I just downloaded the debug build from GH right before testing.

That's a weird commit SHA. It doesn't belong to any commit in this PR. Probably some github internal stuff... I can confirm that it is indeed the artifact that can be downloaded here. 

I do see the button on my device. Maybe try another episode?

tonytamsf commented:
> Hi @tonytamsf
> 3.3.2 (37ef84f). I just downloaded the debug build from GH right before testing.

@keunes I can see the transcript button, and Podcasting 2.0 and Bostagram Ball works for me. Could you test again?  Nothing has changed with the build

keunes commented:
Ok, I identified the issue:
* I imported my database, which already had the known episode. Refreshing the feed after having started streaming the episode: no transcript.
* Import database, refresh feed, then start streaming: transcript shows.
* No db import, clean add feed, start streaming: transcript shows.

Not sure if the problematic scenario is supposed to be supported, but it'd be nice.

> we already have too many toolbar icons for small screens. Having it in the overflow menu is indeed hard to discover

I'm just thinking: is it possible to switch around the share button and the transcript, while keeping the share visible if there's no transcript? (I.e., can we set a maximum number of icons for in the top bar and let Android figure out which one, depending on button relevance?)

Reason:
* we want to maintain the same number of buttons,
* we want to display the share if possible (recently changed to make it always visible),
* but I would like to make transcript more discoverable when relevant, taking priority over share (which is known already)
  * Given the limited number of episodes with a transcript, if the above can be applied I don't expect any comments about share disappearing. Also because most of the episodes which do have a transcript button probably serve a tech audience.


> I wanted to merge your PR before we start working on redesigning the player screen

I fully agree, but can we agree that this should be one of our priorities then? I'm a bit afraid that otherwise most of the podcasting 2.0 stuff will remain with little impact because we can't/don't want to expose them too much.

ByteHamster commented:
> Not sure if the problematic scenario is supposed to be supported, but it'd be nice.

I don't think supporting the edge case of importing an old database into the new app and playing before refreshing is a case we have to spend too much time on. Supporting that would complicate things without a real benefit.

> I'm just thinking: is it possible to switch around the share button and the transcript, while keeping the share visible if there's no transcript? (I.e., can we set a maximum number of icons for in the top bar and let Android figure out which one, depending on button relevance?)

We already let Android figure out the number. If it thinks there is space for the transcript, it already shows it. I don't think it ever does, though. Maybe on tablets.

I'm not a big fan of displaying the share button in different places for different episodes. That feels a bit inconsistent.

keunes commented:
> I'm not a big fan of displaying the share button in different places for different episodes. That feels a bit inconsistent.

Like I said: the general public probably won't be affected. Also, I wouldn't say they are "different places"; the buttons and overflow menu are right next to each other, both in the end corner of the top app bar.

I am really afraid, and I'm sorry to bring this up again, that the player screen redesign won't be prioritised (because it's a big & complex work, which will linger just like multiple queues) and then all the effort of implementing innovative changes in the ecosystem won't reach their potential. As we always say: workarounds are likely to stay. I would prefer to avoid that transcripts fall victim to this, even at the cost of tiny inconsistency for a minor amount of episodes.

ByteHamster commented:
Considering other new features like podcast:socialinteract, we cannot add them all to the toolbar icons. Even if we decided we want to move the "share" button. If you want to make it more prominent, I would rather add it to the buttons below the cover. It does make the cover smaller, but only if chapters or comments are supported for that episode. Next to chapters and description is a more logical place than next to sleep timer and favorite anyway.

Matth7878 commented:
Does having vertical buttons along cover has been considered ? I don't feel it's a good idea because cover won't be centered but it could be a solution to fit more buttons

tonytamsf commented:
I am in favor of merging this PR as is, to get it in the hands of real users during beta. I am pretty certain they will have one to two bugs or feature requests before we go live.
